File: backend/src/routes/patients.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

from src.database import prisma
from src.core.security import get_current_user
from src.core.model_loader import prever_risco_diabetes

router = APIRouter(prefix="/patients", tags=["Pacientes e IA"])

class PatientCreate(BaseModel):
    name: str
    cpf: str
    birth_date: datetime
    
    # Médicos/Clínicos Obrigatórios
    glicemia: float
    pressao: float
    imc: float
    idade: float
    
    # Opcionais
    gestacoes: Optional[float] = None
    espessura_pele: Optional[float] = None
    insulina: Optional[float] = None
    historico_familiar: Optional[float] = None

class PatientUpdate(BaseModel):
    name: Optional[str] = None
    is_active: Optional[bool] = None
    # Adicionar outros se necessário futuramente

@router.post("/")
async def create_patient_with_ai(data: PatientCreate, current_user_id: int = Depends(get_current_user)):
    try:
        # A Mágica Acontece Aqui: Passa os dados pra Inteligência Artificial antes de ir pro banco
        dados = data.dict()
        prob, risco = prever_risco_diabetes(dados)
        
        # Agora sim salva no PostgreSQL com as previsões atreladas
        patient = await prisma.patient.create(
            data={
                "name": data.name,
                "cpf": data.cpf,
                "birth_date": data.birth_date,
                
                "glicemia": data.glicemia,
                "pressao": data.pressao,
                "imc": data.imc,
                "idade": data.idade,
                
                "gestacoes": data.gestacoes,
                "espessura_pele": data.espessura_pele,
                "insulina": data.insulina,
                "historico_familiar": data.historico_familiar,
                
                "predicao_probabilidade": prob,
                "predicao_risco": risco,
                "predicao_data": datetime.utcnow()
            }
        )
        return patient
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/")
async def get_active_patients(current_user_id: int = Depends(get_current_user)):
    return await prisma.patient.find_many(
        where={"is_active": True},
        # Caso queira filtrar pro medico só ver os que ele criou: 
        # where={"is_active": True, "user_id": current_user_id}
    )

@router.get("/{id}")
async def get_patient_by_id(id: int, current_user_id: int = Depends(get_current_user)):
    patient = await prisma.patient.find_unique(where={"id": id})
    if not patient or not patient.is_active:
        raise HTTPException(status_code=404, detail="Paciente não encontrado.")
    return patient

@router.put("/{id}")
async def update_patient(id: int, data: PatientUpdate, current_user_id: int = Depends(get_current_user)):
    patient = await prisma.patient.find_unique(where={"id": id})
    if not patient:
        raise HTTPException(status_code=404, detail="Paciente não existe.")
        
    update_data = data.dict(exclude_unset=True)
    updated = await prisma.patient.update(where={"id": id}, data=update_data)
    return updated

File: backend-js/src/routes/patients.js
const router = require('express').Router();
const prisma = require('../database');
const { preverRiscoDiabetes } = require('../core/modelLoader');

router.post('/', async (req, res) => {
  const {
    name, cpf, birth_date,
    glicemia, pressao, imc, idade,
    gestacoes, espessura_pele, insulina, historico_familiar,
  } = req.body;

  try {
    const { probabilidade, risco } = await preverRiscoDiabetes({
      glicemia, pressao, imc, idade,
      gestacoes, espessura_pele, insulina, historico_familiar,
    });

    const patient = await prisma.patient.create({
      data: {
        name,
        cpf,
        birth_date:             new Date(birth_date),
        glicemia,
        pressao,
        imc,
        idade,
        gestacoes:              gestacoes          ?? null,
        espessura_pele:         espessura_pele     ?? null,
        insulina:               insulina           ?? null,
        historico_familiar:     historico_familiar ?? null,
        predicao_probabilidade: probabilidade,
        predicao_risco:         risco,
        predicao_data:          new Date(),
      },
    });
    res.json(patient);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

router.get('/', async (req, res) => {
  try {
    const patients = await prisma.patient.findMany({ where: { is_active: true } });
    res.json(patients);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

router.get('/:id', async (req, res) => {
  try {
    const patient = await prisma.patient.findUnique({ where: { id: Number(req.params.id) } });
    if (!patient || !patient.is_active)
      return res.status(404).json({ detail: 'Paciente não encontrado.' });
    res.json(patient);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

router.put('/:id', async (req, res) => {
  const { name, is_active } = req.body;
  try {
    const patient = await prisma.patient.findUnique({ where: { id: Number(req.params.id) } });
    if (!patient) return res.status(404).json({ detail: 'Paciente não existe.' });

    const data = {};
    if (name       !== undefined) data.name      = name;
    if (is_active  !== undefined) data.is_active = is_active;

    const updated = await prisma.patient.update({ where: { id: Number(req.params.id) }, data });
    res.json(updated);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

module.exports = router;
